modules/setting/oauth2.go

```diff
@@ -128,28 +128,47 @@ func loadOAuth2From(rootCfg ConfigProvider) {
 		return
 	}
 
-	OAuth2.JWTSecretBase64 = loadSecret(sec, "JWT_SECRET_URI", "JWT_SECRET")
+	jwtSecretBase64 := loadSecret(sec, "JWT_SECRET_URI", "JWT_SECRET")
 
 	if !filepath.IsAbs(OAuth2.JWTSigningPrivateKeyFile) {
 		OAuth2.JWTSigningPrivateKeyFile = filepath.Join(AppDataPath, OAuth2.JWTSigningPrivateKeyFile)
 	}
 
 	if InstallLock {
-		if _, err := generate.DecodeJwtSecretBase64(OAuth2.JWTSecretBase64); err != nil {
-			_, OAuth2.JWTSecretBase64, err = generate.NewJwtSecretWithBase64()
+		jwtSecretBytes, err := generate.DecodeJwtSecretBase64(jwtSecretBase64)
+		if err != nil {
+			jwtSecretBytes, jwtSecretBase64, err = generate.NewJwtSecretWithBase64()
 			if err != nil {
 				log.Fatal("error generating JWT secret: %v", err)
 			}
-
 			saveCfg, err := rootCfg.PrepareSaving()
 			if err != nil {
 				log.Fatal("save oauth2.JWT_SECRET failed: %v", err)
 			}
-			rootCfg.Section("oauth2").Key("JWT_SECRET").SetValue(OAuth2.JWTSecretBase64)
-			saveCfg.Section("oauth2").Key("JWT_SECRET").SetValue(OAuth2.JWTSecretBase64)
+			rootCfg.Section("oauth2").Key("JWT_SECRET").SetValue(jwtSecretBase64)
+			saveCfg.Section("oauth2").Key("JWT_SECRET").SetValue(jwtSecretBase64)
 			if err := saveCfg.Save(); err != nil {
 				log.Fatal("save oauth2.JWT_SECRET failed: %v", err)
 			}
 		}
+		generalSigningSecret.Store(&jwtSecretBytes)
+	}
+}
+
+var generalSigningSecret atomic.Pointer[[]byte]
+
+func GetGeneralTokenSigningSecret() []byte {
+	old := generalSigningSecret.Load()
+	if old == nil || len(*old) == 0 {
+		jwtSecret, _, err := generate.NewJwtSecretWithBase64()
+		if err != nil {
+			log.Fatal("Unable to generate general JWT secret: %s", err.Error())
+		}
+		if generalSigningSecret.CompareAndSwap(old, &jwtSecret) {
```

Comment:
Why do you use `CompareAndSwap` here? `jwtSecret` was generated some lines above and the pointer will always be different to what is stored in `generalSigningSecret`.

Reply:
For concurrency.

For example, two threads call the `GetGeneralTokenSigningSecret`, this CAS makes sure they get the same result.

Reply:
Are you sure? Two threads call that method. Both `Load` and get a nil pointer. Both call `NewJwtSecretWithBase64` and get a new secret. Then the second thread sleeps and the first thread swaps the value and returns. Second thread awakes, swaps the value again because it's different and returns with a different value.

Reply:
>  Second thread awakes, swaps the value again because it's different and returns with a different value.

Why the second thread would swap? The stored value has changed, it doesn't match "old(nil)", then the CAS fails.

Reply:
![image](https://github.com/go-gitea/gitea/assets/2114189/66cd1027-43d3-4559-a7a3-bde126138f2a)

Reply:
> But why not just use a mutex?

Well, IMO "atomic pointer" just matches this case. It doesn't need any Lock/Unlock code and it is easier to read.

In many cases, "atomic pointer" and "sync.Mutex" could all work.

Reply:
Our discussion would be a argument against "easier to read". I think a mutex is more maintainable here. It's not really a hot path to need the possible performance boost.

Reply:
By using "mutex":

* It needs an extra global variable
* It needs to add at lest two pairs of "Lock/Unlock"

I still think CAS is designed for this case. But if most maintainers would like to use mutex approach, it's also fine to me, feel free to edit the PR directly.

Reply:
If you explain it in a codecommit next to the atomic pointer declaration ... i'm fine with it.
It should have the tldr of this conversation

Else i would also go with mutex ...

Reply:
> If you explain it in a codecommit next to the atomic pointer declaration

Sorry I didn't get the point about how to explain. "atomic compare-and-swap" is a basic concurrency programing memory model. Feel free to add more information to help to improve.

> Else i would also go with mutex ...

Do you prefer to introduce an extra variable for "mutex" and consume more memory? IIRC you have proposed some PRs to help to remove global variables to reduce memory usage. 😄

Reply:
Please read the whole text

Reply:
> Please read the whole text

Sorry ... I didn't get the point.

> > If you explain it in a codecommit next to the atomic pointer declaration ... i'm fine with it.
It should have the tldr of this conversation

Do you mean that add a comment into code? Or something else ..... sometimes I have difficulty to read native English 😢

Reply:
ah ok I'll push it to your branch ...

Reply:
> ah ok I'll push it to your branch ...

a9b376ba9ce697538c09c57c7153e396a1acf041

Reply:
PS: `codecommit` is a misspell of mine - should be called "code comment" ... :sweat_smile: 

what I use for the language barrier is https://www.deepl.com/translator - but of course you cant always have it at hand ... it also consumes time :sweat: